### PR TITLE
gangway: BulkJobStatusChange: return early and finish later

### PR DESCRIPTION
**Pull request by Prucek**
This change will accept the `BulkJobStatusChange` request and return immediately and it will finish the processing in a separate goroutine. On bigger prow instances, with a lot of Prowjobs, it was observed, that it takes a lot of time to only to list all Prowjobs, and calls to this endpoint caused: `context timeout`

### CODE REVIEW

**Comment by smg247** [comment]
This takes a bit of a different direction than what I was expecting. My impression is that this service will nearly always be used when there are more than 10 jobs to update. If we have 10 or less, we will likely find it just as easy to just manually update each one. With that context, I think it makes sense to always return the list of jobs to be updated, prior to actually updating any of them. Then spawn a new thread to update all the jobs. What do you think about that approach?

File: pkg/gangway/gangway.go [hunk 1]
```diff
@@ -170,87 +173,96 @@ func TranslateProwJobType(prowJobType prowcrd.ProwJobType) JobExecutionType {
...
+		return &emptypb.Empty{}, status.Error(codes.InvalidArgument, err.Error())
 	}
 
-	var errs []error
-	jobExecutions := []*JobExecution{}
-	for _, pj := range pjList.Items {
-		if !isMatchingCondition(pj, request) {
-			continue
+	// creating a context that does not get cancelled and finish the task in the background
+	context, _ := context.WithTimeout(context.WithoutCancel(ctx), CONTEXT_TIMEOUT)
```
Comment (smg247) [inline pkg/gangway/gangway.go:197]: Can this be initialized closer to where it is used?

File: pkg/gangway/gangway.go [hunk 2]
```diff
@@ -170,87 +173,96 @@ func TranslateProwJobType(prowJobType prowcrd.ProwJobType) JobExecutionType {
...
 
-		pj.Status.State = prowcrd.ProwJobState(strings.ToLower(request.GetJobStatusChange().GetDesired().String()))
-		updatedPj, err := gw.ProwJobClient.Update(ctx, &pj, metav1.UpdateOptions{})
-		if err != nil {
-			logrus.WithError(err).Errorf("failed to update ProwJob status")
-			errs = append(errs, fmt.Errorf("%s, failed to update ProwJob status: %s", err, pj.Name))
-			continue
-		}
-		logrus.WithField("name", pj.Name).Infof("ProwJob status updated to: %s", updatedPj.Status.State)
+	return &emptypb.Empty{}, nil
```
Comment (smg247) [inline pkg/gangway/gangway.go:238]: Do you know what status code this ends up returning?

File: pkg/gangway/gangway.go [hunk 3]
```diff
@@ -170,87 +173,96 @@ func TranslateProwJobType(prowJobType prowcrd.ProwJobType) JobExecutionType {
...
+	}
+	if request.JobType == JobExecutionType_PERIODIC {
+		labelSelector.MatchLabels["prow.k8s.io/type"] = string(prowcrd.PeriodicJob)
 	}
-	jobsAffected := &JobsAffected{
-		Count:         int32(jobLen),
-		JobExecutions: jobExecutions,
+	if request.Refs != nil {
+		if request.Refs.GetOrg() != "" {
+			labelSelector.MatchLabels["prow.k8s.io/refs.org"] = request.Refs.GetOrg()
```
Comment (smg247) [inline pkg/gangway/gangway.go:254]:
```suggestion
		if org := request.Refs.GetOrg(); org != "" {
			labelSelector.MatchLabels["prow.k8s.io/refs.org"] = org
```

File: pkg/gangway/gangway.go [hunk 4]
```diff
@@ -170,87 +173,96 @@ func TranslateProwJobType(prowJobType prowcrd.ProwJobType) JobExecutionType {
...
+		}
+		if request.Refs.GetRepo() != "" {
```
Comment (smg247) [inline pkg/gangway/gangway.go:256]: same as above

File: pkg/gangway/gangway.go [hunk 5]
```diff
@@ -170,87 +173,96 @@ func TranslateProwJobType(prowJobType prowcrd.ProwJobType) JobExecutionType {
...
-		jobExec := &JobExecution{
-			Id:        updatedPj.Name,
-			JobName:   updatedPj.Spec.Job,
-			JobType:   JobExecutionType(TranslateProwJobType(updatedPj.Spec.Type)),
-			JobStatus: TranslateProwJobStatus(&updatedPj.Status),
-		}
-		jobExecutions = append(jobExecutions, jobExec)
+func getRequestLabelSelector(request *BulkJobStatusChangeRequest) *metav1.LabelSelector {
+	labelSelector := &metav1.LabelSelector{MatchLabels: make(map[string]string)}
+	if request.JobType == JobExecutionType_PRESUBMIT {
```
Comment (smg247) [inline pkg/gangway/gangway.go:243]: this could be a switch statement

File: test/integration/test/gangway_test.go [hunk 6]
```diff
@@ -515,24 +516,22 @@ func TestGangwayBulkJobStatusChange(t *testing.T) {
...
 				}
 			} else {
-				if jobsAffected.Count != int32(tt.expectedCount) {
-					t.Fatalf("Expected %d jobs to be affected, got %d", tt.expectedCount, jobsAffected.Count)
-				}
-				for _, job := range jobsAffected.JobExecutions {
-					if job.JobStatus != gangway.JobExecutionStatus_ABORTED {
-						t.Fatalf("Expected job status to be %q, got %q", gangway.JobExecutionStatus_ABORTED, job.JobStatus)
+				for _, jobExecution := range jobExecutions {
+					timeout := 120 * time.Second
```
Comment (smg247) [inline test/integration/test/gangway_test.go:526]: how did you decide on the timeout? that seems like a long time to wait during an integration test, does it actually take that long?

Comment (Prucek) [reply test/integration/test/gangway_test.go:526]: No, it doesn't take long, I copied it from the test above

File: pkg/gangway/gangway.go (hunk 2 above)
Comment (Prucek) [reply pkg/gangway/gangway.go:238]: I have not tried it, because I had to change the proto file, I will check that, but it should be `OK 0`

Comment (Prucek) [reply pkg/gangway/gangway.go:238]: yes, it is so and return this: 
`{}`

File: pkg/gangway/gangway.go [hunk 7]
```diff
@@ -170,87 +173,97 @@ func TranslateProwJobType(prowJobType prowcrd.ProwJobType) JobExecutionType {
...
+			}
+			pj.Status.State = prowcrd.ProwJobState(strings.ToLower(request.GetJobStatusChange().GetDesired().String()))
+			updatedPj, err := gw.ProwJobClient.Update(context, &pj, metav1.UpdateOptions{})
+			if err != nil {
+				logrus.WithError(err).Errorf("failed to update ProwJob status")
+				continue
+			}
+			logrus.WithField("name", pj.Name).Infof("ProwJob status updated to: %s", updatedPj.Status.State)
 		}
+		cancel()
```
Comment (smg247) [inline pkg/gangway/gangway.go:237]: Do we need this? Is it possible to just use `_` instead of creating the `cancel` var?

Comment (Prucek) [reply pkg/gangway/gangway.go:237]: yes, unfortunately, otherwise the linter job is failing

**Review by droslean — APPROVED** [review]

**Pull request merged**